api/app/Http/Controllers/EntityController.php

```diff
@@ -222,14 +237,149 @@ public function deleteOne($id)
+            return $this->getModel()->findByIdentifier($id);
+        } catch (ModelNotFoundException $e) {
+            return $this->getModel()->newInstance();
         }
+    }
 
-        $this->getRepository()->deleteMany($models);
-        return $this->getResponse()->noContent();
+    /**
+     * @param $id
+     * @return BaseModel
+     */
+    protected function findOrFailEntity($id)
+    {
+        $this->validateId($id, $this->getModel()->getKeyName(), $this->getIdValidationRule());
+
+        try {
+            return $this->getModel()->findByIdentifier($id);
+        } catch (ModelNotFoundException $e) {
+            throw $this->notFoundException($this->getModel()->getKeyName());
+        }
+    }
+
+    /**
+     * @return int
+     */
+    protected function countEntities()
+    {
+        return $this->getModel()->count();
+    }
+
+    /**
+     * @param null $limit
+     * @param null $offset
+     * @return Collection
+     */
+    protected function getAllEntities($limit = null, $offset = null)
+    {
+        return $this->getModel()->take($limit)->skip($offset)->get();
+    }
+
+    /**
+     * @param $requestCollection
+     * @return Collection
+     */
+    protected function findOrFailCollection($requestCollection)
+    {
+        $ids = $this->getIds($requestCollection, $this->getModel()->getKeyName(), $this->getIdValidationRule());
+
+        if (!empty($ids)) {
+            $models = $models = $this->getModel()->findMany($ids);
+        } else {
+            throw $this->notFoundManyException($ids, $this->getModel()->newCollection(), $this->getModel()->getKeyName());
+        }
+
+        if ($models && count($ids) !== $models->count()) {
+            throw $this->notFoundManyException($ids, $models, $this->getModel()->getKeyName());
+        }
+
+        return $models;
+    }
+
+    /**
+     * @param $requestCollection
+     * @return Collection
+     */
+    protected function findCollection($requestCollection)
+    {
+        $ids = $this->getIds($requestCollection, $this->getModel()->getKeyName(), $this->getIdValidationRule());
+
+        if (!empty($ids)) {
+            $models = $models = $this->getModel()->findMany($ids);
+        } else {
+            $models = $this->getModel()->newCollection();
+        }
+
+        return $models;
+    }
+
+    /**
+     * @return BaseModel
+     */
+    protected function getModel()
+    {
+        return $this->model;
+    }
+
+    /**
+     * Get id validation rule from model validation rules
+     * Can be overriden by validateIdRule property
+     * @return null|string
+     */
+    protected function getIdValidationRule()
+    {
+        if ($this->validateIdRule) {
+            return $this->validateIdRule;
+        }
+
+        $validationRules = $this->getValidationRules();
+
+        if (isset($validationRules[$this->getModel()->getKeyName()])) {
+            return $validationRules[$this->getModel()->getKeyName()];
+        }
+
+        return null;
+    }
+
```

Comment:
this getIdValidationRule() is hard to be tested
means we are doing something wrong...
can do it via reflection